lib/sshutils/tcpip.go

```diff
@@ -55,3 +61,63 @@ type TCPIPForwardReq struct {
 	// Port is the port to listen on.
 	Port uint32
 }
+
+// ParseTCPIPForwardReq parses an SSH request's payload into a TCPIPForwardReq.
+func ParseTCPIPForwardReq(data []byte) (*TCPIPForwardReq, error) {
+	var r TCPIPForwardReq
+	if err := ssh.Unmarshal(data, &r); err != nil {
+		log.Infof("failed to parse TCP IP Forward request: %v", err)
+		return nil, trace.Wrap(err)
+	}
+	return &r, nil
+}
+
+type channelOpener interface {
+	OpenChannel(name string, data []byte) (ssh.Channel, <-chan *ssh.Request, error)
+}
+
+// StartRemoteListener listens on the given listener and forwards any accepted
+// connections over a new "forwarded-tcpip" channel.
+func StartRemoteListener(ctx context.Context, sshConn channelOpener, srcAddr, dstAddr string, listener net.Listener) error {
+	srcHost, srcPort, err := SplitHostPort(srcAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	dstHost, dstPort, err := SplitHostPort(dstAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	req := ForwardedTCPIPRequest{
+		Addr:     srcHost,
+		Port:     srcPort,
+		OrigAddr: dstHost,
+		OrigPort: dstPort,
+	}
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	reqBytes := ssh.Marshal(req)
+
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				if !utils.IsOKNetworkError(err) {
+					log.WithError(err).Warn("failed to accept connection")
+				}
+				return
+			}
+
+			ch, rch, err := sshConn.OpenChannel(teleport.ChanForwardedTCPIP, reqBytes)
```

Comment:
I don't think so in this case. The addresses here are the ones that we're forwarding from/to, they don't have anything to do with the addresses of the ssh connection itself.

Reply:
https://datatracker.ietf.org/doc/html/rfc4254#section-7.2

```
      byte      SSH_MSG_CHANNEL_OPEN
      string    "forwarded-tcpip"
      uint32    sender channel
      uint32    initial window size
      uint32    maximum packet size
      string    address that was connected
      uint32    port that was connected
      string    originator IP address
      uint32    originator port
```

We should have local address and port and then remote address and port of `conn`, not the original local and remote addresses.

Reply:
If my understanding of the ssh client is correct, it has to receive a) the originator host and port that it initially requested in the initial `tcpip-forward` message, and b) the host and port that it should dial locally. Otherwise, the client won't have enough information to proxy the channel to and from the correct places. I've tried replacing both addresses with the addresses of `conn`, and each combo results in forwarding not working anymore.

Reply:
Are you sure about this? The "address that was connected" and "port that was connected" are the ones from the initial request (not sure about the behavior of 0.0.0.0), but looking at both the openssh source code and the `golang.org/x/crypto/(*ssh.Client).Listen` implementation, the "originator IP address" and "originator port" seem to be the _remote_ address of the incoming connection, which will obviously change on every new incoming connection - what would be the point of adding extra static data that's not necessary to identify the connection, otherwise?

Reply:
OK, I see how that makes sense, but consider the situation where the client has requested more than one forwarded port (i.e. `-R 8000:localhost:8000 -R 8001:localhost:8001`). If the client receives a `forwarded-tcpip` request and the originator address doesn't have either port 8000 or 8001, how can it know which listener to forward the connection to? `forwarded-tcpip` is a channel open request, so it's not connected to a listener to start with. Here in the [go ssh client](https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.21.0:ssh/tcpip.go;drc=b2d7c26edb17864f117d8b0ee73c1843bcc6090f;l=277) it's using the received address to make that decision.

Possible tangent, but it might help our understanding: For the other non-originator address, I think `tsh`/`ssh` remote forwarding is a special case where we could ignore it. The server only ever knows the listen address, so the ssh protocol would allow the server to choose an arbitrary dial address, send the dial address in a `forwarded-tcpip` request, and have the client dispatch the dialing. In our case, we happen to always have a one dial address for each listen address, so that dispatching never comes up.